services/spar/src/Spar/API.hs

```diff
@@ -278,31 +279,47 @@ validateNewIdP ::
   (HasCallStack, m ~ Spar) =>
   SAML.IdPMetadata ->
   TeamId ->
+  Maybe SAML.IdPId ->
   m IdP
-validateNewIdP _idpMetadata _idpExtraInfo = do
+validateNewIdP _idpMetadata teamId mReplaced = do
   _idpId <- SAML.IdPId <$> SAML.createUUID
+  oldIssuers :: [SAML.Issuer] <- case mReplaced of
+    Nothing -> pure []
+    Just replaced -> do
+      idp <- wrapMonadClient (Data.getIdPConfig replaced) >>= maybe (throwSpar SparNotFound) pure
+      pure $ (idp ^. SAML.idpMetadata . SAML.edIssuer) : (idp ^. SAML.idpExtraInfo . wiOldIssuers)
   let requri = _idpMetadata ^. SAML.edRequestURI
+      _idpExtraInfo = WireIdP teamId oldIssuers Nothing
   enforceHttps requri
   wrapMonadClient (Data.getIdPIdByIssuer (_idpMetadata ^. SAML.edIssuer)) >>= \case
     Nothing -> pure ()
     Just _ -> throwSpar SparNewIdPAlreadyInUse
   pure SAML.IdPConfig {..}
 
-idpUpdate :: Maybe UserId -> SAML.IdPId -> IdPMetadataInfo -> Spar IdP
-idpUpdate zusr idpid (IdPMetadataValue raw xml) = idpUpdateXML zusr idpid raw xml
-
-idpUpdateXML :: Maybe UserId -> SAML.IdPId -> Text -> SAML.IdPMetadata -> Spar IdP
-idpUpdateXML zusr idpid raw idpmeta = withDebugLog "idpUpdate" (Just . show . (^. SAML.idpId)) $ do
+-- | FUTUREWORK: 'idpUpdateXML' is only factored out of this function for symmetry with
+-- 'idpCreate', which is not a good reason.  make this one function and pass around
+-- 'IdPMetadataInfo' directly where convenient.
+idpUpdate :: Maybe UserId -> IdPMetadataInfo -> SAML.IdPId -> Spar IdP
+idpUpdate zusr (IdPMetadataValue raw xml) idpid = idpUpdateXML zusr raw xml idpid
+
+-- | The @isPure@ param only has an effect if the issuer changes.  If 'True', a new IdP is
```

Comment:
`isPure` is not a parameter anymore. comment is old

Reply:
8c0c70b6baa38c1f0f3185b21d34b25b0c0617e6